pkg/server/plugin_invoker.go

```diff
@@ -61,51 +67,228 @@ type invokerError struct {
 	message string
 }
 
-// requestReplyHeaders is the set of headers that should be copied from incoming message to outgoing message, if present
-var requestReplyHeaders []string
+var errorType = reflect.TypeOf((*error)(nil)).Elem()
+var nilError reflect.Value
+
+var Trace *log.Logger // exported so the main package can redirect output
 
 func init() {
-	requestReplyHeaders = []string{CorrelationId}
+	var e error
+	nilError = reflect.ValueOf(&e).Elem()
+
+	Trace = log.New(os.Stdout, "", log.LstdFlags)
 }
 
-func (pi *pluginInvoker) invoke(in *function.Message) *function.Message {
-	arg, err := pi.messageToFunctionArgs(in)
-	if err != nil {
-		return propagateHeaders(in, errorToMessage(err))
+// type shared captures all coordination state between the two goroutines and the Call()
+// function, to make function signatures more digestible.
+type shared struct {
+	input  reflect.Value // reflects the 'input' channel to the user function
+	output reflect.Value // reflects the 'output' channel of the user function
+	fnErrs reflect.Value // reflects the 'errors' channel of the user function (optional)
+
+	sidecar function.MessageFunction_CallServer
+
+	errs chan error    // used to signal errors to the Call() function
+	done chan struct{} // used to broadcast early cancellation to all parties, and opt out of an otherwise blocking channel operation
+
+	// TODO: make Accept passing a responsibility of the sidecar
+	// TODO: make correlationId propagation a responsibility of the sidecar
+	acceptC chan []string
+	// Temp: assume it's the invoker responsibility to propagate the correlationId when present
+	// this will only make sense with f: X -> Y functions
+	correlationIdC chan []string
+}
+
+func (pi *pluginInvoker) Call(callServer function.MessageFunction_CallServer) error {
+	
+	input := makeChannel(pi.inType)
+	channelValues := pi.fn.Call([]reflect.Value{input})
+
+	ss := &shared{
+		input:          input,
+		output:         channelValues[0],
+		sidecar:        callServer,
+		errs:           make(chan error, 1),
+		done:           make(chan struct{}),
+		acceptC:        make(chan []string, 1),
+		correlationIdC: make(chan []string, 1),
+	}
+
+	if len(channelValues) == 2 {
+		ss.fnErrs = channelValues[1]
 	}
-	reflectiveArgs := [] reflect.Value{reflect.ValueOf(arg)}
 
-	reflectiveResult := pi.fn.Call(reflectiveArgs)
+	// Sidecar => function input
+	go pi.sidecar2Function()(ss)
 
-	invocationResult := make([]interface{}, len(reflectiveResult))
-	for i, v := range reflectiveResult {
-		invocationResult[i] = v.Interface()
+	// Function output => sidecar
+	go pi.function2Sidecar()(ss)
+
+	var err error
+	for i := 0; i < 2+(len(channelValues)-1); i++ { // Read errors from both goroutines + 1 optional from fn itself
```

Comment:
I thought errors from the fn itself would appear on `ss.fnErrs` rather than `ss.errs`.